elephant/joint_isi_dithering_class.py

```diff
@@ -0,0 +1,438 @@
+"""
+Module to generate surrogates of spike trains by using the joint-ISI dithering.
+Features are provided to separate the preprocessing from the main process.
+
+Original implementation by: Peter Bouss [p.bouss@fz-juelich.de]
+:copyright: Copyright 2019 by the Elephant team, see `doc/authors.rst`.
+:license: Modified BSD, see LICENSE.txt for details.
+"""
+
+import numpy as np
+import quantities as pq
+import neo
+
+import elephant.spike_train_surrogates as surr
+import elephant.statistics as stats
+from scipy.ndimage import gaussian_filter
+
+
+class JointISISpace:
+    """
+    The class :class:`Joint_ISI_Space` is implemented for Joint-ISI dithering
+    as a continuation of the ideas of Louis et al. (2010) and Gerstein (2004).
+
+    When creating an class instance all necessary preprocessing steps are done,
+    to use the method dithering().
+
+    To make the dithering-procedure stable, the spiketrain needs to pass two
+    thresholds, that the Joint-ISI dithering is applied, if not uniform
+    dithering is used.
+
+    Attributes
+    ----------
+    st: neo.SpikeTrain
+        For this spiketrain the surrogates will be created
+
+    n_surr: int, optional
+        Number of surrogates to be created.
+        Default: 1
+    dither: pq.Quantity
+        The range of the dithering for the uniform dithering,
+        which is also used for the method 'window'.
+        Default: 15.*pq.ms
+    unit: pq.unit
```

Comment:
Is changing the units of the output spiketrain an important feature to include in this function? It seems to me that in 99.9% of cases you would expect the same units as the input spike train, and else, one would rescale the result of the jisi dither.